File: docs/breez-sdk/snippets/kotlin_mpp_lib/shared/src/commonMain/kotlin/com/example/kotlinmpplib/IssuingTokens.kt
package com.example.kotlinmpplib

import breez_sdk_spark.*
import com.ionspin.kotlin.bignum.integer.BigInteger

class IssuingTokens {
    fun getTokenIssuer(sdk: BreezSdk): TokenIssuer {
        // ANCHOR: get-token-issuer
        val tokenIssuer = sdk.getTokenIssuer()
        // ANCHOR_END: get-token-issuer
        return tokenIssuer
    }

    suspend fun createToken(tokenIssuer: TokenIssuer) {
        // ANCHOR: create-token
        try {
            val request =
                    CreateIssuerTokenRequest(
                            name = "My Token",
                            ticker = "MTK",
                            decimals = 6.toUInt(),
                            isFreezable = false,
                            maxSupply = BigInteger.fromLong(1_000_000L)
                    )
            val tokenMetadata = tokenIssuer.createIssuerToken(request)
            // Log.v("Breez", "Token identifier: ${tokenMetadata.identifier}")
        } catch (e: Exception) {
            // Handle exception
        }
        // ANCHOR_END: create-token
    }

    suspend fun createTokenWithCustomAccountNumber() {
        // ANCHOR: custom-account-number
        val accountNumber = 21u

        val mnemonic = "<mnemonic words>"
        val seed = Seed.Mnemonic(mnemonic, null)
        val config = defaultConfig(Network.MAINNET)
        config.apiKey = "<breez api key>"

        try {
            val builder = SdkBuilder(config, seed)
            builder.withDefaultStorage("./.data")

            // Set the account number for the SDK
            val keySetConfig = KeySetConfig(
                keySetType = KeySetType.DEFAULT,
                useAddressIndex = false,
                accountNumber = accountNumber
            )
            builder.withKeySet(keySetConfig)

            val sdk = builder.build()
        } catch (e: Exception) {
            // handle error
        }
        // ANCHOR_END: custom-account-number
    }

    suspend fun mintToken(tokenIssuer: TokenIssuer) {
        // ANCHOR: mint-token
        try {
            val request =
                    MintIssuerTokenRequest(
                            amount = BigInteger.fromLong(1_000L),
                    )
            val payment = tokenIssuer.mintIssuerToken(request)
        } catch (e: Exception) {
            // Handle exception
        }
        // ANCHOR_END: mint-token
    }

    suspend fun burnToken(tokenIssuer: TokenIssuer) {
        // ANCHOR: burn-token
        try {
            val request =
                    BurnIssuerTokenRequest(
                            amount = BigInteger.fromLong(1_000L),
                    )
            val payment = tokenIssuer.burnIssuerToken(request)
        } catch (e: Exception) {
            // Handle exception
        }
        // ANCHOR_END: burn-token
    }

    suspend fun getTokenMetadata(tokenIssuer: TokenIssuer) {
        // ANCHOR: get-token-metadata
        try {
            val tokenBalance = tokenIssuer.getIssuerTokenBalance()
            // Log.v("Breez", "Token balance: ${tokenBalance.balance}")

            val tokenMetadata = tokenIssuer.getIssuerTokenMetadata()
            // Log.v("Breez", "Token ticker: ${tokenMetadata.ticker}")
        } catch (e: Exception) {
            // Handle exception
        }
        // ANCHOR_END: get-token-metadata
    }

    suspend fun freezeToken(tokenIssuer: TokenIssuer) {
        // ANCHOR: freeze-token
        try {
            val sparkAddress = "<spark address>"
            // Freeze the tokens held at the specified Spark address
            val freezeRequest =
                    FreezeIssuerTokenRequest(
                            address = sparkAddress,
                    )
            val freezeResponse = tokenIssuer.freezeIssuerToken(freezeRequest)

            // Unfreeze the tokens held at the specified Spark address
            val unfreezeRequest =
                    UnfreezeIssuerTokenRequest(
                            address = sparkAddress,
                    )
            val unfreezeResponse = tokenIssuer.unfreezeIssuerToken(unfreezeRequest)
        } catch (e: Exception) {
            // Handle exception
        }
        // ANCHOR_END: freeze-token
    }
}

File: docs/breez-sdk/snippets/wasm/issuing_tokens.ts
import {
  type Payment,
  type TokenMetadata,
  type BreezSdk,
  type TokenIssuer,
  type Seed,
  defaultConfig,
  SdkBuilder,
  KeySetType
} from '@breeztech/breez-sdk-spark'
import { Network } from 'node:inspector'

// Init stub
const init = async () => { }

const getTokenIssuer = (sdk: BreezSdk) => {
  // ANCHOR: get-token-issuer
  const tokenIssuer = sdk.getTokenIssuer()
  // ANCHOR_END: get-token-issuer
  return tokenIssuer
}

const createToken = async (tokenIssuer: TokenIssuer): Promise<TokenMetadata> => {
  // ANCHOR: create-token
  const tokenMetadata = await tokenIssuer.createIssuerToken({
    name: 'My Token',
    ticker: 'MTK',
    decimals: 6,
    isFreezable: false,
    maxSupply: BigInt(1_000_000)
  })
  console.debug(`Token identifier: ${tokenMetadata.identifier}`)
  // ANCHOR_END: create-token
  return tokenMetadata
}

const createTokenWithCustomAccountNumber = async () => {
  // ANCHOR: custom-account-number
  await init()

  const accountNumber = 21

  const mnemonic = '<mnemonic words>'
  const seed: Seed = { type: 'mnemonic', mnemonic, passphrase: undefined }
  const config = defaultConfig('mainnet')
  config.apiKey = '<breez api key>'
  let builder = SdkBuilder.new(config, seed)
  builder = await builder.withDefaultStorage('./.data')

  // Set the account number for the SDK
  builder = builder.withKeySet({
    keySetType: 'default',
    useAddressIndex: false,
    accountNumber
  })

  const sdk = await builder.build()
  // ANCHOR_END: custom-account-number
}

const mintToken = async (tokenIssuer: TokenIssuer): Promise<Payment> => {
  // ANCHOR: mint-token
  const payment = await tokenIssuer.mintIssuerToken({
    amount: BigInt(1_000)
  })
  // ANCHOR_END: mint-token
  return payment
}

const burnToken = async (tokenIssuer: TokenIssuer): Promise<Payment> => {
  // ANCHOR: burn-token
  const payment = await tokenIssuer.burnIssuerToken({
    amount: BigInt(1_000)
  })
  // ANCHOR_END: burn-token
  return payment
}

const getTokenMetadata = async (tokenIssuer: TokenIssuer): Promise<TokenMetadata> => {
  // ANCHOR: get-token-metadata
  const tokenBalance = await tokenIssuer.getIssuerTokenBalance()
  console.debug(`Token balance: ${tokenBalance.balance}`)

  const tokenMetadata = await tokenIssuer.getIssuerTokenMetadata()
  console.debug(`Token ticker: ${tokenMetadata.ticker}`)
  // ANCHOR_END: get-token-metadata
  return tokenMetadata
}

const freezeToken = async (tokenIssuer: TokenIssuer): Promise<void> => {
  // ANCHOR: freeze-token
  const sparkAddress = '<spark address>'
  // Freeze the tokens held at the specified Spark address
  const freezeResponse = await tokenIssuer.freezeIssuerToken({
    address: sparkAddress
  })

  // Unfreeze the tokens held at the specified Spark address
  const unfreezeResponse = await tokenIssuer.unfreezeIssuerToken({
    address: sparkAddress
  })
  // ANCHOR_END: freeze-token
}
